components/brave_extension/extension/brave_extension/background/api/cosmeticFilterAPI.ts

```diff
@@ -38,36 +39,54 @@ export const injectClassIdStylesheet = (tabId: number, classes: string[], ids: s
   })
 }
 
-// Fires on content-script loaded
-export const applyAdblockCosmeticFilters = (tabId: number, frameId: number, url: string, hide1pContent: boolean) => {
+export const getAdblockCosmeticFiltersKey = (tabId: number, frameId: number) => {
+  return `adblockCosmeticFilter-${tabId}-${frameId}`
+}
+
+export const fetchAdblockCosmeticFilters = (url: string, tabId: number, frameId: number, hide1pContent: boolean) => {
   chrome.braveShields.urlCosmeticResources(url, async (resources) => {
     if (chrome.runtime.lastError) {
+      shieldsPanelActions.cosmeticFilterResourcesReady(tabId, frameId, hide1pContent, undefined)
       console.warn('Unable to get cosmetic filter data for the current host', chrome.runtime.lastError)
       return
     }
 
-    if (frameId === 0) {
-      if (hide1pContent) {
-        resources.force_hide_selectors.push(...resources.hide_selectors)
-      } else {
-        informTabOfCosmeticRulesToConsider(tabId, resources.hide_selectors)
-      }
+    shieldsPanelActions.cosmeticFilterResourcesReady(tabId, frameId, hide1pContent, resources)
+  })
+}
 
-      let styledStylesheet = ''
-      if (resources.force_hide_selectors.length > 0) {
-        styledStylesheet += resources.force_hide_selectors.join(',') + '{display:none!important;}\n'
-      }
-      for (const selector in resources.style_selectors) {
-        styledStylesheet += selector + '{' + resources.style_selectors[selector].join(';') + ';}\n'
-      }
-      chrome.tabs.insertCSS(tabId, {
-        code: styledStylesheet,
-        cssOrigin: 'user',
-        runAt: 'document_start'
-      })
+export const applyAdblockCosmeticFiltersWithResources = (tabId: number, frameId: number, hide1pContent: boolean, resources: UrlCosmeticResourcesType) => {
+  if (frameId === 0) {
+    if (hide1pContent) {
+      resources.force_hide_selectors.push(...resources.hide_selectors)
+    } else {
+      informTabOfCosmeticRulesToConsider(tabId, resources.hide_selectors)
     }
 
-    shieldsPanelActions.cosmeticFilterRuleExceptions(tabId, frameId, resources.exceptions, resources.injected_script || '', resources.generichide)
+    let styledStylesheet = ''
+    if (resources.force_hide_selectors.length > 0) {
+      styledStylesheet += resources.force_hide_selectors.join(',') + '{display:none!important;}\n'
+    }
+    for (const selector in resources.style_selectors) {
+      styledStylesheet += selector + '{' + resources.style_selectors[selector].join(';') + ';}\n'
+    }
+    chrome.tabs.insertCSS(tabId, {
+      code: styledStylesheet,
+      cssOrigin: 'user',
+      runAt: 'document_start'
+    })
+  }
+  setTimeout(() => shieldsPanelActions.cosmeticFilterRuleExceptions(tabId, frameId, resources.exceptions, resources.injected_script || '', resources.generichide), 0)
```

Comment:
This `setTimeout` is used to avoid nested redux call. However, this adds another overhead.

Reply:
I thought the general idea is to _remove_ all cosmetic filter actions from redux reducer? And then just have:

```
events -> data retreival -> content_script
```

If that's beyond the scope of this PR then that's absolutely fine and we can do the simplification in a follow-up.

Reply:
Yes, let's do more cleanup/refactoring in a f/u :)